orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py

```diff
@@ -0,0 +1,654 @@
+
+        @staticmethod
+        def backward(ctx, grad_output):
+            # Bypass the reduce as if we are using only 1 GPU.
+            return grad_output
+
+    class MegatronFModel(torch.nn.Module):
+        def __init__(self, output_size):
+            super(MegatronFModel, self).__init__()
+            self.copy_ = MegatronFFunction.apply
+            self.bias = Parameter(torch.empty(
+                output_size,
+                device=torch.cuda.current_device(),
+                dtype=torch.float))
+
+            # Always initialize bias to zero.
+            with torch.no_grad():
+                self.bias.uniform_()
+
+        def forward(self, model_input):
+            model_input = model_input + self.bias
+            out = self.copy_(model_input)
+            return out
+
+    output_size = 1024
+
+    def model_builder():
+        return MegatronFModel(output_size)
+
+    def input_generator():
+        return torch.randn(output_size, dtype=torch.float)
+
+    # generate a label that have same shape as forward output.
+    label_input = torch.ones([output_size])
+
+    run_training_test_and_compare(model_builder, input_generator, label_input)
+
+
+def test_ScalarAndTuple():
+    class ScalarAndTupleFunction(torch.autograd.Function):
+        @staticmethod
+        def forward(ctx, input, alpha, beta, gamma):
+            ctx.save_for_backward(input)
+            ctx.alpha = alpha
+            ctx.beta = beta
+            ctx.gamma = gamma
+            return alpha * beta[0] * beta[1] * gamma * input.clamp(min=0)
+
+        @staticmethod
+        def backward(ctx, grad_output):
+            input, = ctx.saved_tensors
+            alpha = ctx.alpha
+            beta = ctx.beta
+            gamma = ctx.gamma
+            grad_input = grad_output.clone()
+            grad_input[input < 0] = 0
+            return alpha * beta[0] * beta[1] * gamma * grad_input, None, None, None
+
+    class ScalarAndTupleModel(torch.nn.Module):
+        def __init__(self, output_size):
+            super(ScalarAndTupleModel, self).__init__()
+            self.activation = ScalarAndTupleFunction.apply
+            self.linear_a = torch.nn.Linear(output_size, output_size)
+            self.linear_b = torch.nn.Linear(output_size, output_size)
+
+        def forward(self, x):
+            h = self.linear_a(x)
+            h = self.activation(h, 5.0, (-1.0, 2.0), -1.0)
+            h = self.linear_b(h)
+            return h
+
+    output_size = 2
+
+    def model_builder():
+        return ScalarAndTupleModel(output_size)
+
+    def input_generator():
+        return torch.randn(output_size, dtype=torch.float)
+
+    # generate a label that have same shape as forward output.
+    label_input = torch.ones([output_size])
+
+    run_training_test_and_compare(model_builder, input_generator, label_input)
+
+
+@pytest.mark.skip(reason="This test is not correct. All tensors modified by in-place operattions should be mark_dirty(...).")
```

Comment:
What is the intention behind skipping these tests?

Reply:
The tests are runnable but accordingly to Pytorch document, they are not leagl use of some functions.